packages/jsx-email/src/renderer/raw.ts

```diff
@@ -1,12 +1,94 @@
 const START_TAG = '__COMMENT_START';
 const END_TAG = '__COMMENT_END';
+
+// Note: opaque wrapper tokens used to safely carry raw payloads through
+// rehype/HTML parsing without introducing nested HTML comments.
+export const RAW_WRAPPER_START = '__JSX_EMAIL_RAW_START__';
+export const RAW_WRAPPER_END = '__JSX_EMAIL_RAW_END__';
+
+const escapeRe = (s: string) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+
+// Exported regular expressions for Raw wrapper handling. Keeping these as
+// named exports makes their behavior explicit and testable.
+export const RAW_MARKER_CONTENT_RE = new RegExp(
+  `${escapeRe(RAW_WRAPPER_START)}([\\s\\S]*?)${escapeRe(RAW_WRAPPER_END)}`
+);
+
+// Allow an optional stray "--" to precede the closing angle bracket in the
+// closing tag. Some HTML serializers can emit a comment terminator sequence
+// immediately before our synthetic element closer when the payload ends with
+// "-->", resulting in `</jsx-email-raw-->`. Matching this variant makes the
+// unwrap tolerant without impacting standard cases.
+export const RAW_ELEMENT_MARKER_UNWRAP_RE = new RegExp(
+  `<jsx-email-raw[^>]*?>\\s*${escapeRe(RAW_WRAPPER_START)}([\\s\\S]*?)${escapeRe(
+    RAW_WRAPPER_END
+  )}\\s*<\\/jsx-email-raw(?:--)?\\s*>(?:<\\/jsx-email-raw\\s*>)?`,
+  'g'
+);
+
+export const RAW_ELEMENT_LEGACY_UNWRAP_RE = new RegExp(
+  `<jsx-email-raw[^>]*?>\\s*<!--([\\s\\S]*?)-->\\s*<\\/jsx-email-raw(?:--)?\\s*>(?:<\\/jsx-email-raw\\s*>)?`,
+  'g'
+);
 export function escapeForRawComponent(input: string): string {
-  // escape comment sequences
-  return input.replace(/<!--/g, START_TAG).replace(/-->/g, END_TAG);
+  // Escape comment sequences and guard against user content accidentally
+  // containing our wrapper tokens.
+  return input
+    .replace(/<!--/g, START_TAG)
+    .replace(/-->/g, END_TAG)
+    .replace(new RegExp(RAW_WRAPPER_START, 'g'), '__RAW_EMBEDDED_START__')
+    .replace(new RegExp(RAW_WRAPPER_END, 'g'), '__RAW_EMBEDDED_END__');
+}
+
+// Minimal HTML entity decoding to restore characters that rehype stringifies
+// inside text nodes. We intentionally support the common named entities and
+// numeric (decimal and hex) forms used by the serializer.
+function decodeHtmlEntities(value: string): string {
```

Comment:
this seems brittle and unnecessary. if rehype is encoding characters in text nodes and that's corrupting the output and causing snapshot failures, when there's a greater issue at play, because this was not a problem with the original implementation. rehype is the source of truth on correct behavior. we use things like comments to work within it's rulesets, without having to bend those rules.